File: OnTheAlgo/AlgoImpl/GolangImpl/impl/labExercise/week2/sameFirstLetter.go
package impl

import "strings"

/*

Given a list of non-empty strings, write a method that returns a Map<String, String> with a key for every different first letter seen, with the value of all the strings starting with that letter appended with a comma (,) together in the order they appear in the list.

Test case 1:
sameFirstLetter(["alice", "bob", "apple", "banana"]) →  {"a": "alice,apple", "b": "bob,banana"}

Test case 2:
sameFirstLetter(["after", "all", "this", "time", "always"]) →  {"a":"after,all,always", "t": "this,time"}

*/
func sameFirstLetter(arr []string) map[string]string {
	m := map[string]string{}
	for i := 0; i < len(arr); i++ {
		word := arr[i]
		// v := m[string(word[1])]
		k := string(word[0:1])
		m[k] += word + ","
	}

	for k, v := range m {
		m[k] = strings.Trim(v, ",")
	}

	return m
}

File: OnTheCode/Java/Advanced-OO-Programming/src/lab2/SameFirstLetter/SameFirstLetter.java
package lab2.SameFirstLetter;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SameFirstLetter {

    /**
     * Create a map with first letter as key and words with that same
     * first letter separated by comma.
     * For example, numWords(["alice", "bob", "apple", "banana"]) →
     * {"a": "alice,apple", "b": "bob,banana"}.
     * @param list is a list of strings.
     * The strings are non-empty.
     * @return a map with first letter and comma-separated-words pair.
     */
    public static Map<String, String> sameFirstLetter(List<String> list) {
        Map<String, String> map = new HashMap<>();

        for (String word : list) {
            String key = String.valueOf(word.charAt(0));
            if (!map.containsKey(key)) {
                map.put(key, word);
            } else {
                String value = map.get(key);
                value += "," + word;
                map.put(key, value);
            }
        }
		
		return map;
    }

}